### PR TITLE
[Google Services] Major Cleanup

### PR DESCRIPTION
**Pull request by StevenRoddis**
...e

Should I have written `<test>`s for this?

### CODE REVIEW

File: src/chrome/content/rules/GoogleServices.xml [hunk 1]
```diff
@@ -234,7 +233,7 @@ Fetch error: http://googleusercontent.com/ => http://googleusercontent.com/: (6,
 	<target host="*.googlesource.com" />
 	<target host="*.googlesyndication.com" />
 	<target host="www.googletagservices.com" />
-	<target host="googleusercontent.com" />
```
Comment (jsha) [inline src/chrome/content/rules/GoogleServices.xml]: Please also remove the corresponding `<rule>` below.

**Comment by jsha** [comment]
Yep, this ruleset is going to need a lot of `<test>` URLs. I know it's a huge amount of work, but especially for a big ruleset like this one it's important to have a lot of tests. It might be worth checking out the `exrex` Python module I mention in issue #1069.

Note that I had temporarily disabled ruleset coverage checking and forgot to turn it back on, which is why you haven't been seeing coverage errors for your branches so far.

If you have the time to add the necessary coverage, that would be terrific. I would suggest taking the opportunity to simplify this ruleset a lot. This ruleset has been around for a long time, and in the meanwhile Google has greatly improved their HTTPS infrastructure and made HTTPS the default for a lot of situations. So for instance, the super long regex that lists the paths that work under HTTPS can probably be simplified to match just on hostnames. I am happy to assume that any path under www.google.* works on HTTPS today. If users report issues, we can add exceptions for specific paths.

**Comment by StevenRoddis** [comment]
I'm still working on this, just wanted to push my changes to let others view them and see the build status.

**Comment by jsha** [comment]
Looking good, thanks so much for working on this.

By the way, if you'd like to be able to run the tests locally, and you have a Debian or Ubuntu machine, you can run `./install-dev-dependencies.sh`, which should get you all the prerequisites needed to be able to run `./test-ruleset-coverage.sh` locally. It's also fine to keep pushing changes to see the status, but this might save you some time since Travis takes ~3 minutes to run.

**Comment by StevenRoddis** [comment]
Not quite sure how to debug the error it's giving me.

**Comment by jsha** [comment]
Yeah, the Relax-NG checker doesn't give very useful error messages, I'm afraid. I think we may be able to add friendly messages to the XML validator syntax.

The nice thing is that it *does* provide line numbers. Line 69 in Google_subdomains.xml is `%lt;target host="client*.google.com" /&gt;`. Target tags only accept wildcards for a whole DNS label.

**Comment by StevenRoddis** [comment]
>Target tags only accept wildcards for a whole DNS label

Ah, that makes sense.

**Comment by StevenRoddis** [comment]
What is the best way to run `https-everywhere-checker` on these rulesets to double check my work?

**Comment by jsha** [comment]
For future reference, the easy way to run the ruleset checker for a single rule is:

python2.7 https-everywhere-checker/src/https_everywhere_checker/check_rules.py https-everywhere-checker/manual.checker.config src/chrome/content/rules/RULE_NAME

It looks like the tests are failing because of unrelated changes checked in recently, I will fix.

**Comment by StevenRoddis** [comment]
@jsha thank you

**Comment by StevenRoddis** [comment]
hm.. I don't know why it's giving errors for `IEEE.xml` I haven't touched that file.
`git diff --name-only 4f8fef18a609bb268a6a6490377c301fb6a9796e -- src`
src/chrome/content/rules/Google-mismatches.xml
src/chrome/content/rules/Google.com_Subdomains.xml
src/chrome/content/rules/Google.com_Subdomains_Complex.xml
src/chrome/content/rules/Google.org.xml
src/chrome/content/rules/Google.tld_Subdomains.xml
src/chrome/content/rules/Google.xml
src/chrome/content/rules/GoogleAPIs.xml
src/chrome/content/rules/GoogleCanada.xml
src/chrome/content/rules/GoogleImages.xml
src/chrome/content/rules/GoogleSearch.xml
src/chrome/content/rules/GoogleServices.xml
src/chrome/content/rules/GoogleServices_Complex.xml

**Comment by StevenRoddis** [comment]
I ran `git diff --name-only 4f8fef18a609bb268a6a6490377c301fb6a9796e -- src | xargs ls | xargs check-https-rules ~/manual.conf`
https://gist.github.com/StevenRoddis/1674af6fa8cb74781b94

I'm a bit confused with some of the output:

>2015-03-01 17:26:58,720 ERROR src/chrome/content/rules/GoogleAPIs.xml: Fetch error: http://api.recaptcha.net/ => https://www.google.com/recaptcha/api/: None [build/bdist.linux-x86_64/egg/https_everywhere_checker/check_rules.py:92]

http://api.recaptcha.net does work just 404s on /

The majority of `(200)` are actually `HTTP/1.1 301 Moved Permanently`

**Comment by StevenRoddis** [comment]
hm... might need some help getting the checker to work.

`# git diff --name-only 28ec030ccc6ef4f3482b439ffbd152b6972b6a49 -- src | xargs ls | xargs python2.7 ~/https-everywhere-checker/src/https_everywhere_checker/check_rules.py ~/manual.conf``

>ls: cannot access src/chrome/content/rules/Droplr.com.xml: No such file or directory
>Exception in thread Thread-5 (most likely raised during interpreter shutdown):
>Traceback (most recent call last):
>  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
>  File "/root/https-everywhere-checker/src/https_everywhere_checker/check_rules.py", line 81, in run
>  File "/usr/lib/python2.7/Queue.py", line 168, in get
>  File "/usr/lib/python2.7/threading.py", line 333, in wait
>`<type 'exceptions.TypeError'>`: 'NoneType' object is not callable

**Comment by jsha** [comment]
That is odd, I'll take a look.

**Comment by jsha** [comment]
It looks like you have a different version of https-everywhere-checker than I've been using. I recommend using it via `git submodule init` / `git submodule update`. The version currently used in HTTPS Everywhere is different than the current master in hiviah's repo.

Here are the current errors I get from this branch:

```
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.ca/doubleclick (200) => https://www.google.com/doubleclick (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com/doubleclick (200) => https://www.google.com/doubleclick (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com.au/doubleclick (200) => https://www.google.com/doubleclick (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.ca/help (200) => https://www.google.com/help (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com/help (200) => https://www.google.com/help (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com.au/help (200) => https://www.google.com/help (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.ca/pacman (200) => https://www.google.com/pacman (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com/pacman (200) => https://www.google.com/pacman (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com.au/pacman (200) => https://www.google.com/pacman (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.ca/postini (200) => https://www.google.com/postini (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com/postini (200) => https://www.google.com/postini (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com.au/postini (200) => https://www.google.com/postini (404)
ERROR src/chrome/content/rules/Google.xml: Fetch error: http://www.google.ca/favicon.ico => https://www.google.ca/favicon.ico: 'NoneType' object has no attribute 'xpath'
ERROR src/chrome/content/rules/Google.xml: Fetch error: http://www.google.com/favicon.ico => https://www.google.com/favicon.ico: 'NoneType' object has no attribute 'xpath'
ERROR src/chrome/content/rules/Google.xml: Fetch error: http://www.google.com.au/favicon.ico => https://www.google.com.au/favicon.ico: 'NoneType' object has no attribute 'xpath'
ERROR src/chrome/content/rules/Google.xml: Fetch error: http://www.google.ca/mapmaker => https://www.google.ca/mapmaker: (28, 'Resolving timed out after 10516 milliseconds')
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.ca/mobile (200) => https://www.google.ca/mobile (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com/mobile (200) => https://www.google.com/mobile (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com.au/mobile (200) => https://www.google.com.au/mobile (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://google.com/ig (200) => https://www.google.com/ig (404)
ERROR src/chrome/content/rules/Google.xml: Non-2xx HTTP code: http://www.google.com/ig (200) => https://www.google.com/ig (404)
```

Most of those are legit. I'll look into the 'NoneType' errors.

**Comment by StevenRoddis** [comment]
>'NoneType' object has no attribute 'xpath'

Perhaps because it's an image file not html?

**Comment by jsha** [comment]
I'm pretty sure that's not it. The tool doesn't try to parse the returned HTML at all.

**Comment by jsha** [comment]
Other than the favicon thing (which I'll treat as spurious for now), what's the status of this branch? If it's close to ready I'd like to merge it for the upcoming release.

**Comment by StevenRoddis** [comment]
It's at the review stage, I just need to review my rules. It's very close to being ready, few days?

**Comment by StevenRoddis** [comment]
All my testing has been good. What the timeline for the 5.x (dev) releases?
I'd say this is mergable.

**Comment by jsha** [comment]
Great, thanks! It looks like like there are merge conflicts. Can you merge the latest master into your branch and then I will merge it?

**Comment by StevenRoddis** [comment]
https://github.com/EFForg/https-everywhere/commit/c2ac2f0553f06281800c1ca5f9d5395d8e6f8268
@2d1 Just a head sup I removed your comment from `GoogleServices.xml` as it doesn't house any google.com domains anymore.

**Comment by StevenRoddis** [comment]
Interesting message:
`src/chrome/content/rules/GoogleMaps.xml failed XML validity: Double hyphen within comment: <!--
Ignore this error if it ever happens again, it's , line 5, column 1`

I didn't know this wasn't allowed in XML. It's easy to remove:

```xml
<!--
Ignore this error if it ever happens again, it's ok. Please reenable this ruleset if it happens again.
Disabled by https-everywhere-checker because:
Non-2xx HTTP code: http://maps-api-ssl.google.com/ (200) => https://maps-api-ssl.google.com/ (404)
------

	Problematic domains:
```

**Comment by jsha** [comment]
FWIW, I was totally wrong about the 'xpath' error message. The checker does in fact parse the returned documents, to calculate Levenshtein distance of the tree structures. I'm working on a fix to the checker.

There are still a few test URLs you added that get 200 without rewrite and 404 with rewrite. I'm fixing the relevant rule, then I'll manually merge.

File: https-everywhere-checker [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit 2ec94f1ce11a2046408a8df228447e2f0dd85cda
```
Comment (jsha) [inline https-everywhere-checker]: Why this diff?

File: src/chrome/content/rules/GoogleSorry.xml [hunk 3]
```diff
@@ -1,7 +1,15 @@
-<ruleset name="GoogleSorry">
-  <target host="sorry.google.com" />
-  <target host="www.google.com" />
-  <target host="google.com" />
+<ruleset name="Google Sorry">
```
Comment (jsha) [inline src/chrome/content/rules/GoogleSorry.xml]: I'm pretty sure this rule can be eliminated:

```
$ curl -si www.google.com/sorry | grep Location
Location: http://ipv4.google.com/sorry/IndexRedirect?continue=http://www.google.com/sorry
$ curl -si https://www.google.com/sorry | grep Location
Location: https://ipv4.google.com/sorry/IndexRedirect?continue=https://www.google.com/sorry

Then we can just put sorry.google.com as a simple rule in Google.xml.

File: https-everywhere-checker (hunk 2 above)
Comment (jsha) [inline https-everywhere-checker]: FWIW, I'm guessing that using an incorrect version of https-everywhere-checker is why you are getting Travis test failures.

**Pull request merged**